CI failures hardly related, but still pretty interesting:


> Stress test (debug) — Check timeout expired [Details](https://s3.amazonaws.com/clickhouse-test-reports/56996/51eec7a185e0ebd15363b4ceb3a981903815ebf1/stress_test__debug_.html)

```
2023.12.19 15:57:34.457840 [ 547 ] {} <Fatal> Application: Child process was terminated by signal 9 (KILL). If it is not done by 'forcestop' command or manually, the possible cause is OOM Killer (see 'dmesg' and look at the '/var/log/kern.log' for the details).
```

> Stress test (tsan) — Hung check failed, possible deadlock found (see hung_check.log) [Details](https://s3.amazonaws.com/clickhouse-test-reports/56996/51eec7a185e0ebd15363b4ceb3a981903815ebf1/stress_test__tsan_.html)

```
2023-12-19 16:48:05,220 Checking if some queries hung
Using queries from '/usr/share/clickhouse-test/queries' directory
Connecting to ClickHouse server... OK
Received exception from server (version 23.12.1):
Code: 219. DB::Exception: Received from localhost:9000. DB::Exception: New table appeared in database being dropped or detached. Try again.. (DATABASE_NOT_EMPTY)
(query: DETACH DATABASE db01802)
Traceback (most recent call last):
  File "/usr/bin/stress", line 374, in <module>
    main()
  File "/usr/bin/stress", line 359, in main
    res = call(cmd, shell=True, stdout=tee.stdin, stderr=STDOUT, timeout=600)
  File "/usr/lib/python3.10/subprocess.py", line 347, in call
    return p.wait(timeout=timeout)
  File "/usr/lib/python3.10/subprocess.py", line 1209, in wait
    return self._wait(timeout=timeout)
  File "/usr/lib/python3.10/subprocess.py", line 1951, in _wait
    raise TimeoutExpired(self.args, timeout)
subprocess.TimeoutExpired: Command '/usr/bin/clickhouse-test --client-option max_untracked_memory=1Gi max_memory_usage_for_user=0 memory_profiler_step=1Gi --database=system --hung-check --report-logs-stats 00001_select_1' timed out after 600 seconds

```

> Upgrade check (msan) — check_status.tsv doesn't exists [Details](https://s3.amazonaws.com/clickhouse-test-reports/56996/51eec7a185e0ebd15363b4ceb3a981903815ebf1/upgrade_check__msan_.html)

<details>
<summary>MemorySanitizer: use-of-uninitialized-value</summary>


```
==231025==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x5646cfdc5fdd in DB::ReplicatedMergeTreeAttachThread::run() build_docker/./src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.cpp:83:9
    #1 0x5646cfdcfc7e in DB::ReplicatedMergeTreeAttachThread::ReplicatedMergeTreeAttachThread(DB::StorageReplicatedMergeTree&)::$_0::operator()() const build_docker/./src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.cpp:24:82
    #2 0x5646cfdcfc7e in decltype(std::declval<DB::ReplicatedMergeTreeAttachThread::ReplicatedMergeTreeAttachThread(DB::StorageReplicatedMergeTree&)::$_0&>()()) std::__1::__invoke[abi:v15000]<DB::ReplicatedMergeTreeAttachThread::ReplicatedMergeTreeAttachThread(DB::StorageReplicatedMergeTree&)::$_0&>(DB::ReplicatedMergeTreeAttachThread::ReplicatedMergeTreeAttachThread(DB::StorageReplicatedMergeTree&)::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23
    #3 0x5646cfdcfc7e in void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::ReplicatedMergeTreeAttachThread::ReplicatedMergeTreeAttachThread(DB::StorageReplicatedMergeTree&)::$_0&>(DB::ReplicatedMergeTreeAttachThread::ReplicatedMergeTreeAttachThread(DB::StorageReplicatedMergeTree&)::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9
...

Member fields were destroyed
    #0 0x56469d600d1d in __sanitizer_dtor_callback_fields (/usr/bin/clickhouse+0x7c86d1d) (BuildId: b02ef26d094f99621fd60a2d973ca865dffde192)
    #1 0x5646cfdc4177 in DB::ReplicatedMergeTreeAttachThread::~ReplicatedMergeTreeAttachThread() build_docker/./src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.h:37:20
    #2 0x5646cfdc4177 in DB::ReplicatedMergeTreeAttachThread::~ReplicatedMergeTreeAttachThread() build_docker/./src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.cpp:32:1
    #3 0x5646ce6f1e5d in std::__1::__optional_destruct_base<DB::ReplicatedMergeTreeAttachThread, false>::~__optional_destruct_base[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/optional:261:21
    #4 0x5646ce6f1e5d in DB::StorageReplicatedMergeTree::~StorageReplicatedMergeTree() build_docker/./src/Storages/StorageReplicatedMergeTree.cpp:5267:1
...
```

https://s3.amazonaws.com/clickhouse-test-reports/56996/51eec7a185e0ebd15363b4ceb3a981903815ebf1/upgrade_check__msan_/stderr.log



</details>

> Upgrade check (tsan) — check_status.tsv doesn't exists [Details](https://s3.amazonaws.com/clickhouse-test-reports/56996/51eec7a185e0ebd15363b4ceb3a981903815ebf1/upgrade_check__tsan_.html)


<details>
<summary>ThreadSanitizer: data race</summary>


```
WARNING: ThreadSanitizer: data race (pid=1018668)
  Read of size 8 at 0x7b7800c6ea28 by main thread:
    #0 std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo>::get[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:801:16 (clickhouse+0x194f74c8) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #1 std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo>::operator bool[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:833:16 (clickhouse+0x194f74c8)
    #2 bool std::__1::operator!=[abi:v15000]<DB::BackgroundSchedulePoolTaskInfo>(std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo> const&, std::nullptr_t) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:1254:30 (clickhouse+0x194f74c8)
    #3 DB::BackgroundSchedulePoolTaskHolder::operator bool() const build_docker/./src/Core/BackgroundSchedulePool.h:173:55 (clickhouse+0x194f74c8)
    #4 DB::BackgroundJobsAssignee::finish() build_docker/./src/Storages/MergeTree/BackgroundJobsAssignee.cpp:111:9 (clickhouse+0x194f74c8)
    #5 DB::StorageMergeTree::shutdown(bool) build_docker/./src/Storages/StorageMergeTree.cpp:190:36 (clickhouse+0x19a6e56b) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #6 DB::IStorage::flushAndShutdown(bool) build_docker/./src/Storages/IStorage.h:573:9 (clickhouse+0x1726ed4d) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #7 DB::DatabaseWithOwnTablesBase::shutdown() build_docker/./src/Databases/DatabasesCommon.cpp:311:20 (clickhouse+0x1726ed4d)
    #8 DB::DatabaseOnDisk::shutdown() build_docker/./src/Databases/DatabaseOnDisk.cpp:169:32 (clickhouse+0x171f83cf) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #9 DB::DatabaseCatalog::shutdownImpl() build_docker/./src/Interpreters/DatabaseCatalog.cpp:265:26 (clickhouse+0x175a0e20) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #10 DB::DatabaseCatalog::shutdown() build_docker/./src/Interpreters/DatabaseCatalog.cpp:863:27 (clickhouse+0x175a9dab) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #11 DB::ContextSharedPart::shutdown() build_docker/./src/Interpreters/Context.cpp:567:9 (clickhouse+0x174f7495) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #12 DB::Context::shutdown() build_docker/./src/Interpreters/Context.cpp:4172:13 (clickhouse+0x174e6721) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
...

Previous write of size 8 at 0x7b7800c6ea28 by thread T795 (mutexes: write M0):
    #0 std::__1::enable_if<is_move_constructible<DB::BackgroundSchedulePoolTaskInfo*>::value && is_move_assignable<DB::BackgroundSchedulePoolTaskInfo*>::value, void>::type std::__1::swap[abi:v15000]<DB::BackgroundSchedulePoolTaskInfo*>(DB::BackgroundSchedulePoolTaskInfo*&, DB::BackgroundSchedulePoolTaskInfo*&) build_docker/./contrib/llvm-project/libcxx/include/__utility/swap.h:37:7 (clickhouse+0x194f71a8) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #1 std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo>::swap[abi:v15000](std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo>&) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:761:9 (clickhouse+0x194f71a8)
    #2 std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo>::operator=[abi:v15000](std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo>&&) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:723:38 (clickhouse+0x194f71a8)
    #3 DB::BackgroundSchedulePoolTaskHolder::operator=(DB::BackgroundSchedulePoolTaskHolder&&) build_docker/./src/Core/BackgroundSchedulePool.h:165:110 (clickhouse+0x194f71a8)
    #4 DB::BackgroundJobsAssignee::start() build_docker/./src/Storages/MergeTree/BackgroundJobsAssignee.cpp:103:16 (clickhouse+0x194f71a8)
    #5 DB::StorageMergeTree::startup() build_docker/./src/Storages/StorageMergeTree.cpp:151:40 (clickhouse+0x19a6e293) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #6 DB::DatabaseOrdinary::startupTableAsync(DB::AsyncLoader&, std::__1::unordered_set<std::__1::shared_ptr<DB::LoadJob>, std::__1::hash<std::__1::shared_ptr<DB::LoadJob>>, std::__1::equal_to<std::__1::shared_ptr<DB::LoadJob>>, std::__1::allocator<std::__1::shared_ptr<DB::LoadJob>>>, DB::QualifiedTableName const&, DB::LoadingStrictnessLevel)::$_0::operator()(DB::AsyncLoader&, std::__1::shared_ptr<DB::LoadJob> const&) const build_docker/./src/Databases/DatabaseOrdinary.cpp:212:24 (clickhouse+0x17222fd2) (BuildId: eb6f1638a1dca515120703137057aa43dcfd8685)
    #7 decltype(std::declval<DB::DatabaseOrdinary::startupTableAsync
...

```

https://s3.amazonaws.com/clickhouse-test-reports/56996/51eec7a185e0ebd15363b4ceb3a981903815ebf1/upgrade_check__tsan_/stderr.log

</details>

--- a/src/Interpreters/GraceHashJoin.cpp
+++ b/src/Interpreters/GraceHashJoin.cpp
@@ -424,8 +424,10 @@ void GraceHashJoin::initialize(const Block & sample_block)
 {
     left_sample_block = sample_block.cloneEmpty();
     output_sample_block = left_sample_block.cloneEmpty();
-    ExtraBlockPtr not_processed;
+    ExtraBlockPtr not_processed = nullptr;
     hash_join->joinBlock(output_sample_block, not_processed);
+    if (not_processed)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Unhandled not processed block in GraceHashJoin");
     initBuckets();
 }
 
@@ -447,9 +449,6 @@ void GraceHashJoin::joinBlock(Block & block, std::shared_ptr<ExtraBlock> & not_p
     block = std::move(blocks[current_bucket->idx]);
 
     hash_join->joinBlock(block, not_processed);
-    if (not_processed)
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "Unhandled not processed block in GraceHashJoin");
-
     flushBlocksToBuckets<JoinTableSide::Left>(blocks, buckets);
 }
 
@@ -528,6 +527,13 @@ public:
 
     Block nextImpl() override
     {
+        if (not_processed)
+        {
+            Block block = not_processed->block;
+            hash_join->joinBlock(block, not_processed);
+            return block;
+        }
+
         Block block;
         size_t num_buckets = buckets.size();
         size_t current_idx = buckets[current_bucket]->idx;
@@ -565,12 +571,7 @@ public:
             }
         } while (block.rows() == 0);
 
-        ExtraBlockPtr not_processed;
         hash_join->joinBlock(block, not_processed);
-
-        if (not_processed)
-            throw Exception(ErrorCodes::LOGICAL_ERROR, "Unsupported hash join type");
-
         return block;
     }
 
@@ -582,6 +583,8 @@ public:
 
     Names left_key_names;
     Names right_key_names;
+
+    ExtraBlockPtr not_processed = nullptr;
 };
 
 IBlocksStreamPtr GraceHashJoin::getDelayedBlocks()

--- a/src/Interpreters/HashJoin.cpp
+++ b/src/Interpreters/HashJoin.cpp
@@ -1165,9 +1165,11 @@ public:
 
     std::vector<JoinOnKeyColumns> join_on_keys;
 
+    size_t max_joined_block_rows = 0;
     size_t rows_to_add;
     std::unique_ptr<IColumn::Offsets> offsets_to_replicate;
     bool need_filter = false;
+    IColumn::Filter filter;
 
 private:
     std::vector<TypeAndName> type_name;
@@ -1356,7 +1358,7 @@ void setUsed(IColumn::Filter & filter [[maybe_unused]], size_t pos [[maybe_unuse
 /// Joins right table columns which indexes are present in right_indexes using specified map.
 /// Makes filter (1 if row presented in right table) and returns offsets to replicate (for ALL JOINS).
 template <JoinKind KIND, JoinStrictness STRICTNESS, typename KeyGetter, typename Map, bool need_filter, bool multiple_disjuncts>
-NO_INLINE IColumn::Filter joinRightColumns(
+NO_INLINE size_t joinRightColumns(
     std::vector<KeyGetter> && key_getter_vector,
     const std::vector<const Map *> & mapv,
     AddedColumns & added_columns,
@@ -1365,9 +1367,8 @@ NO_INLINE IColumn::Filter joinRightColumns(
     constexpr JoinFeatures<KIND, STRICTNESS> join_features;
 
     size_t rows = added_columns.rows_to_add;
-    IColumn::Filter filter;
     if constexpr (need_filter)
-        filter = IColumn::Filter(rows, 0);
+        added_columns.filter = IColumn::Filter(rows, 0);
 
     Arena pool;
 
@@ -1375,9 +1376,20 @@ NO_INLINE IColumn::Filter joinRightColumns(
         added_columns.offsets_to_replicate = std::make_unique<IColumn::Offsets>(rows);
 
     IColumn::Offset current_offset = 0;
-
-    for (size_t i = 0; i < rows; ++i)
+    size_t max_joined_block_rows = added_columns.max_joined_block_rows;
+    size_t i = 0;
+    for (; i < rows; ++i)
     {
+        if constexpr (join_features.need_replication)
+        {
+            if (unlikely(current_offset > max_joined_block_rows))
+            {
+                added_columns.offsets_to_replicate->resize_assume_reserved(i);
+                added_columns.filter.resize_assume_reserved(i);
+                break;
+            }
+        }
+
         bool right_row_found = false;
 
         KnownRowsHolder<multiple_disjuncts> known_rows;
@@ -1402,7 +1414,7 @@ NO_INLINE IColumn::Filter joinRightColumns(
                     auto row_ref = mapped->findAsof(left_asof_key, i);
                     if (row_ref.block)
                     {
-                        setUsed<need_filter>(filter, i);
+                        setUsed<need_filter>(added_columns.filter, i);
                         if constexpr (multiple_disjuncts)
                             used_flags.template setUsed<join_features.need_flags, multiple_disjuncts>(row_ref.block, row_ref.row_num, 0);
                         else
@@ -1415,7 +1427,7 @@ NO_INLINE IColumn::Filter joinRightColumns(
                 }
                 else if constexpr (join_features.is_all_join)
                 {
-                    setUsed<need_filter>(filter, i);
+                    setUsed<need_filter>(added_columns.filter, i);
                     used_flags.template setUsed<join_features.need_flags, multiple_disjuncts>(find_result);
                     auto used_flags_opt = join_features.need_flags ? &used_flags : nullptr;
                     addFoundRowAll<Map, join_features.add_missing>(mapped, added_columns, current_offset, known_rows, used_flags_opt);
@@ -1427,7 +1439,7 @@ NO_INLINE IColumn::Filter joinRightColumns(
                     if (used_once)
                     {
                         auto used_flags_opt = join_features.need_flags ? &used_flags : nullptr;
-                        setUsed<need_filter>(filter, i);
+                        setUsed<need_filter>(added_columns.filter, i);
                         addFoundRowAll<Map, join_features.add_missing>(mapped, added_columns, current_offset, known_rows, used_flags_opt);
                     }
                 }
@@ -1438,7 +1450,7 @@ NO_INLINE IColumn::Filter joinRightColumns(
                     /// Use first appeared left key only
                     if (used_once)
                     {
-                        setUsed<need_filter>(filter, i);
+                        setUsed<need_filter>(added_columns.filter, i);
                         added_columns.appendFromBlock<join_features.add_missing>(*mapped.block, mapped.row_num);
                     }
 
@@ -1455,7 +1467,7 @@ NO_INLINE IColumn::Filter joinRightColumns(
                 }
                 else /// ANY LEFT, SEMI LEFT, old ANY (RightAny)
                 {
-                    setUsed<need_filter>(filter, i);
+                    setUsed<need_filter>(added_columns.filter, i);
                     used_flags.template setUsed<join_features.need_flags, multiple_disjuncts>(find_result);
                     added_columns.appendFromBlock<join_features.add_missing>(*mapped.block, mapped.row_num);
 
@@ -1470,7 +1482,7 @@ NO_INLINE IColumn::Filter joinRightColumns(
         if (!right_row_found)
         {
             if constexpr (join_features.is_anti_join && join_features.left)
-                setUsed<need_filter>(filter, i);
+                setUsed<need_filter>(added_columns.filter, i);
             addNotFoundRow<join_features.add_missing, join_features.need_replication>(added_columns, current_offset);
         }
 
@@ -1481,11 +1493,11 @@ NO_INLINE IColumn::Filter joinRightColumns(
     }
 
     added_columns.applyLazyDefaults();
-    return filter;
+    return i;
 }
 
 template <JoinKind KIND, JoinStrictness STRICTNESS, typename KeyGetter, typename Map, bool need_filter>
-IColumn::Filter joinRightColumnsSwitchMultipleDisjuncts(
+size_t joinRightColumnsSwitchMultipleDisjuncts(
     std::vector<KeyGetter> && key_getter_vector,
     const std::vector<const Map *> & mapv,
     AddedColumns & added_columns,
@@ -1497,7 +1509,7 @@ IColumn::Filter joinRightColumnsSwitchMultipleDisjuncts(
 }
 
 template <JoinKind KIND, JoinStrictness STRICTNESS, typename KeyGetter, typename Map>
-IColumn::Filter joinRightColumnsSwitchNullability(
+size_t joinRightColumnsSwitchNullability(
     std::vector<KeyGetter> && key_getter_vector,
     const std::vector<const Map *> & mapv,
     AddedColumns & added_columns,
@@ -1514,7 +1526,7 @@ IColumn::Filter joinRightColumnsSwitchNullability(
 }
 
 template <JoinKind KIND, JoinStrictness STRICTNESS, typename Maps>
-IColumn::Filter switchJoinRightColumns(
+size_t switchJoinRightColumns(
     const std::vector<const Maps *> & mapv,
     AddedColumns & added_columns,
     HashJoin::Type type,
@@ -1597,10 +1609,27 @@ ColumnWithTypeAndName copyLeftKeyColumnToRight(
     return right_column;
 }
 
+/// Cut first num_rows rows from block in place and returns block with remaining rows
+Block sliceBlock(Block & block, size_t num_rows)
+{
+    size_t total_rows = block.rows();
+    if (num_rows >= total_rows)
+        return {};
+
+    Block remaining_block = block.cloneEmpty();
+    for (size_t i = 0; i < block.columns(); ++i)
+    {
+        auto & col = block.getByPosition(i);
+        remaining_block.getByPosition(i).column = col.column->cut(num_rows, total_rows - num_rows);
+        col.column = col.column->cut(0, num_rows);
+    }
+    return remaining_block;
+}
+
 } /// nameless
 
 template <JoinKind KIND, JoinStrictness STRICTNESS, typename Maps>
-void HashJoin::joinBlockImpl(
+Block HashJoin::joinBlockImpl(
     Block & block,
     const Block & block_with_columns_to_add,
     const std::vector<const Maps *> & maps_,
@@ -1642,8 +1671,12 @@ void HashJoin::joinBlockImpl(
 
     bool has_required_right_keys = (required_right_keys.columns() != 0);
     added_columns.need_filter = join_features.need_filter || has_required_right_keys;
+    added_columns.max_joined_block_rows = table_join->maxJoinedBlockRows();
+    if (!added_columns.max_joined_block_rows)
+        added_columns.max_joined_block_rows = std::numeric_limits<size_t>::max();
 
-    IColumn::Filter row_filter = switchJoinRightColumns<KIND, STRICTNESS>(maps_, added_columns, data->type, used_flags);
+    size_t num_joined = switchJoinRightColumns<KIND, STRICTNESS>(maps_, added_columns, data->type, used_flags);
+    Block remaining_block = sliceBlock(block, num_joined);
 
     for (size_t i = 0; i < added_columns.size(); ++i)
         block.insert(added_columns.moveColumn(i));
@@ -1654,7 +1687,7 @@ void HashJoin::joinBlockImpl(
     {
         /// If ANY INNER | RIGHT JOIN - filter all the columns except the new ones.
         for (size_t i = 0; i < existing_columns; ++i)
-            block.safeGetByPosition(i).column = block.safeGetByPosition(i).column->filter(row_filter, -1);
+            block.safeGetByPosition(i).column = block.safeGetByPosition(i).column->filter(added_columns.filter, -1);
 
         /// Add join key columns from right block if needed using value from left table because of equality
         for (size_t i = 0; i < required_right_keys.columns(); ++i)
@@ -1688,7 +1721,7 @@ void HashJoin::joinBlockImpl(
                     continue;
 
                 const auto & left_column = block.getByName(required_right_keys_sources[i]);
-                auto right_col = copyLeftKeyColumnToRight(right_key.type, right_col_name, left_column, &row_filter);
+                auto right_col = copyLeftKeyColumnToRight(right_key.type, right_col_name, left_column, &added_columns.filter);
                 block.insert(std::move(right_col));
 
                 if constexpr (join_features.need_replication)
@@ -1709,6 +1742,8 @@ void HashJoin::joinBlockImpl(
         for (size_t pos : right_keys_to_replicate)
             block.safeGetByPosition(pos).column = block.safeGetByPosition(pos).column->replicate(*offsets_to_replicate);
     }
+
+    return remaining_block;
 }
 
 void HashJoin::joinBlockImplCross(Block & block, ExtraBlockPtr & not_processed) const
@@ -1885,7 +1920,11 @@ void HashJoin::joinBlock(Block & block, ExtraBlockPtr & not_processed)
 
         if (joinDispatch(kind, strictness, maps_vector, [&](auto kind_, auto strictness_, auto & maps_vector_)
         {
-            joinBlockImpl<kind_, strictness_>(block, sample_block_with_columns_to_add, maps_vector_);
+            Block remaining_block = joinBlockImpl<kind_, strictness_>(block, sample_block_with_columns_to_add, maps_vector_);
+            if (remaining_block.rows())
+                not_processed = std::make_shared<ExtraBlock>(ExtraBlock{std::move(remaining_block)});
+            else
+                not_processed.reset();
         }))
         {
             /// Joined

--- a/src/Interpreters/HashJoin.h
+++ b/src/Interpreters/HashJoin.h
@@ -447,7 +447,7 @@ private:
     void initRightBlockStructure(Block & saved_block_sample);
 
     template <JoinKind KIND, JoinStrictness STRICTNESS, typename Maps>
-    void joinBlockImpl(
+    Block joinBlockImpl(
         Block & block,
         const Block & block_with_columns_to_add,
         const std::vector<const Maps *> & maps_,

--- a/tests/performance/hashjoin_with_large_output.xml
+++ b/tests/performance/hashjoin_with_large_output.xml
@@ -1,0 +1,64 @@
+<test>
+    <settings>
+        <max_threads>16</max_threads>
+        <max_memory_usage>10G</max_memory_usage>
+    </settings>
+
+    <substitutions>
+        <substitution>
+            <name>settings</name>
+            <values>
+                <value>join_algorithm='hash'</value>
+                <value>join_algorithm='grace_hash'</value>
+            </values>
+        </substitution>
+    </substitutions>
+
+    <create_query>
+        create table test_left
+        (
+        k1 String,
+        v1 String
+        )
+        engine = Memory();
+    </create_query>
+    <create_query>
+        create table test_right
+        (
+        k1 String,
+        v1 String,
+        v2 String,
+        v3 String,
+        v4 String,
+        v5 String,
+        v6 String,
+        v7 String,
+        v8 String,
+        v9 String
+        )
+        engine = Memory();
+    </create_query>
+    <fill_query>insert into test_left SELECT toString(number % 20), toString(number) from system.numbers limit 10000;</fill_query>
+    <fill_query>
+        insert into test_right
+        SELECT
+        toString(number % 20),
+        toString(number * 10000),
+        toString(number * 10000 + 1),
+        toString(number * 10000 + 2),
+        toString(number * 10000 + 3),
+        toString(number * 10000 + 4),
+        toString(number * 10000 + 5),
+        toString(number * 10000 + 6),
+        toString(number * 10000 + 7),
+        toString(number * 10000 + 8)
+        from system.numbers limit 10000;
+    </fill_query>
+
+    <query>
+        select * from test_left all inner join test_right on test_left.k1 = test_right.k1  SETTINGS {settings} format Null
+    </query>
+
+    <drop_query>DROP TABLE IF EXISTS test_left</drop_query>
+    <drop_query>DROP TABLE IF EXISTS test_right</drop_query>
+</test>

